### ISSUE TITLE
💡 [REQUEST] - Read trimming for variable amplicon sizes

### ISSUE THREAD

**Issue by marchoeppner** [bio-raum/FooDMe2#81]
### Is your feature request related to a problem?

Read trimming will yield sub-optiomal results when the amplicon size fluctuates between "smaller than the read length" and "larger than the read length". 

### Describe the solution you'd like

We need an implementation of cutadapt that can combine results from --cutadapt_trim_3p and without in case where the read length is not obviously longer or shorter than the amplicon(s). 

### Description

A process is needed that runs cutadapt iteratively with 3'-5' trimming and without and then combines both results into a non-redundant set.

**Comment by marchoeppner** [comment]
Worse, I think ;)

We are using the pipeline implementation already, since the more "modern" double-trimming idea didn't reliably remove adapters at both ends of a read. 

But for this, we bascially need to run the --cutadapt_trim_3p pipe solution, catch the untrimmed reads in each step, combine them and then run a "normal" trimming on those reads ; and then combine the trimmed reads from all stages into one R1 and R2 file. It's horrible :D

**Comment by gregdenay** [comment]
Can't we just trim once on 5', discard reads that weren't trimmed and trim on 3', keeping all (trimmed and untrimmed)?

**Comment by marchoeppner** [comment]
I was under the impression that "just" requesting 3' and 5' trimming would not guaranteed that *both* types are applied to the same read; which is why we have this clumsy iterative trimming approach for "--cutadapt_trim_3". 

But I could be wrong, might require a bit of testing.

**Comment by gregdenay** [comment]
Wouldn't this work?

```groovy
if (params.cutadapt_trim_3p) {
        //  Trimming on both side
        """
        cutadapt $read_config \\
            --discard-untrimmed \\
            --cores $task.cpus \\
            $args \\
            $reads \\
            $options_5p \\
            --json=${prefix}_forward.json \\
        | cutadapt $read_config \\
            --discard-untrimmed \\
            $args \\
            --cores $task.cpus \\
            $trimmed \\
            $options_3p \\
            --json=${prefix}_reverse.json \\
            -Z - \\
            > ${prefix}.cutadapt.log

        cutadapt_sum_json.py --sample ${meta.sample_id} \\
        --forward ${prefix}_forward.json \\
        --reverse ${prefix}_reverse.json \\
        --output ${meta.sample_id}.cutadapt_mqc.json
        
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            cutadapt: \$(cutadapt --version)
        END_VERSIONS
        """
    } else if (params.cutadapt_trim_flex) {
        //  Trimming on 5' and maybe 3'
        """
        cutadapt $read_config \\
            --discard-untrimmed \\
            --cores $task.cpus \\
            $args \\
            $reads \\
            $options_5p \\
            --json=${prefix}_forward.json \\
        | cutadapt $read_config \\
  # Keep reads that were not trimmed on the 3' side         --discard-untrimmed \\
            $args \\
            --cores $task.cpus \\
            $trimmed \\
            $options_3p \\
            --json=${prefix}_reverse.json \\
            -Z - \\
            > ${prefix}.cutadapt.log

        cutadapt_sum_json.py --sample ${meta.sample_id} \\
        --forward ${prefix}_forward.json \\
        --reverse ${prefix}_reverse.json \\
        --output ${meta.sample_id}.cutadapt_mqc.json
        
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            cutadapt: \$(cutadapt --version)
        END_VERSIONS
        """ 
    } else {
       // Trimming only on 5'
        """
        cutadapt \\
            --discard-untrimmed \\
            -Z \\
            --cores $task.cpus \\
            $args \\
            $trimmed \\
            $reads \\
            $options_5p \\
            --json=${meta.sample_id}_forward.json \\
            > ${prefix}.cutadapt.log

        cutadapt_sum_json.py --sample ${meta.sample_id} \\
        --forward ${prefix}_forward.json \\
        --output ${meta.sample_id}.cutadapt_mqc.json

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            cutadapt: \$(cutadapt --version)
        END_VERSIONS
        """
    }
```

**Comment by gregdenay** [comment]
Implemented in [e65cde0](https://github.com/bio-raum/FooDMe2/commit/e65cde053baabef0518183ed2802378b92874e9e#diff-bd4d55f724d0dee35d7bd37a05518fa14a6962c9092cb33f0c5bb3751e8a2892)
